### PR TITLE
Rebuilt test suite to only test public api

### PR DESCRIPTION
**Pull request by mattpocock**
This PR removes all unit tests from the project and replaces them with a much simpler, more comprehensive test suite.

The test suite now runs the _actual_ codegen tool on the 'examples' folder, and typechecks them. This means:

1. Issues can now be added to our integration tests by adding failing machines to our 'examples' folder. Once this PR is merged I'll add cases for #23 and #21. This is far easier and more transparent than contributing to unit tests.
2. Big changes to internals can be made (such as the upcoming migration away from rollup to pure TS) without any changes to the test suite.
3. No more having to run `yarn local-link` to test local repositories - just copy them and their imported files to the 'examples' folder and run `yarn test:watch`.

### CODE REVIEW

File: .github/workflows/lint.yml [hunk 1]
```diff
@@ -12,7 +12,7 @@ jobs:
         run: yarn install
 
       - name: Typecheck
-        run: yarn x tsc
+        run: yarn x build
```
Comment (Andarist) [inline .github/workflows/lint.yml:15]: `yarn x`? what does it do? can't find any docs for it

File: packages/xstate-compiled/package.json [hunk 2]
```diff
@@ -32,21 +32,26 @@
...
+    "nodemon": "2.0.4"
   },
   "scripts": {
     "local-link": "yarn unlink && npm run build && npm run chmod:index && yarn link",
     "build": "rm -rf bin && tsc && cp -r src/templates bin/templates && cp -r src/fake_node_modules examples/node_modules",
     "prepare": "npm run build",
     "prepublishOnly": "npm run test && npm run build",
-    "test": "jest ./src",
+    "test": "node test.js",
+    "test:watch": "nodemon test.js",
```
Comment (Andarist) [inline packages/xstate-compiled/package.json:47]: How failures look like in the watch mode in the terminal?

Comment (mattpocock) [reply packages/xstate-compiled/package.json:47]: Similar to jest, it visibly fails but doesn't stop the watch.

File: .github/workflows/lint.yml (hunk 1 above)
Comment (mattpocock) [reply .github/workflows/lint.yml:15]: In the root package.json, I have aliased `yarn workspace xstate-codegen` to `yarn x` for ease of use of scripts from the root.

**Review by Andarist — APPROVED** [review]
Just some more thoughts:
1. I'm not entirely saying that unit tests are bad - they certainly have their place in the world. It's just a very small place in my opinion :trollface: For instance, if we get some really crucial and convoluted logic with lots of edge cases in someplace then it might be worthwhile to unit test such, but alternatives should be evaluated first.
2. I very much like the e2e focus here because ultimately we want to provide a good type-checking experience, so we should test the improved type-checking - we just can't assess that from snapshots. While in the future we want to stabilize the produced output because probably we might need to provide some type helpers etc and then this becomes part of the public API as well. For now, though it's not that important because we want to iterate fast. We want to introduce snapshot tests quite soon though as the produced output gives a nice glimpse into how this works and they really might come in handy for code reviews. This is IMHO the prime use case for snapshots - testing generated outputs. All Babel, Rollup, CSS-in-JS libraries and more do this and it's a perfect match for the job. It's just that snapshots should not be the primary way of testing things.
3. I think that in the future we might want to introduce Jest back and isolate test cases. Selective testing is quite important (at least for me) - it would be great to filter which tests one would like to run at a particular moment during development/debugging. It's also easier to focus on the generated output per test than looking through the output generated for many, many tests (that I hope we'll have 😅 ). Jest's filtering capabilities seem to be well-suited for this.

**Comment by mattpocock** [comment]
@Andarist agree on all the above. We have a rather lovely situation in that we get to do e2e testing without any of the usual downsides - slow tests and flakiness. This should cover the vast majority of our bases while also providing a really nice local dev environment for contributors.
